dynd/include/type_functions.hpp

```diff
@@ -96,14 +219,140 @@ PYDYND_API dynd::ndt::type dynd_make_struct_type(PyObject *field_types,
 PYDYND_API dynd::ndt::type
 dynd_make_fixed_dim_type(PyObject *shape, const dynd::ndt::type &element_tp);
 
-/**
- * Implementation of __getitem__ for the wrapped dynd type object.
- */
-PYDYND_API dynd::ndt::type _type_getitem(const dynd::ndt::type &d,
-                                         PyObject *subscript);
+inline dynd::string_encoding_t encoding_from_pyobject(PyObject *encoding_obj)
+{
+  // Default is utf-8
+  if (encoding_obj == Py_None) {
+    return dynd::string_encoding_utf_8;
+  }
+
+  dynd::string_encoding_t encoding = dynd::string_encoding_invalid;
+  std::string encoding_str = pystring_as_string(encoding_obj);
+  switch (encoding_str.size()) {
+  case 4:
+    switch (encoding_str[3]) {
+    case '2':
+      if (encoding_str == "ucs2") {
+        encoding = dynd::string_encoding_ucs_2;
+      }
+      break;
+    case '8':
+      if (encoding_str == "utf8") {
+        encoding = dynd::string_encoding_utf_8;
+      }
+      break;
+    }
+  case 5:
+    switch (encoding_str[1]) {
+    case 'c':
+      if (encoding_str == "ucs_2" || encoding_str == "ucs-2") {
+        encoding = dynd::string_encoding_ucs_2;
+      }
+      break;
+    case 's':
+      if (encoding_str == "ascii") {
+        encoding = dynd::string_encoding_ascii;
+      }
+      break;
+    case 't':
+      if (encoding_str == "utf16") {
+        encoding = dynd::string_encoding_utf_16;
+      }
+      else if (encoding_str == "utf32") {
+        encoding = dynd::string_encoding_utf_32;
+      }
+      else if (encoding_str == "utf_8" || encoding_str == "utf-8") {
+        encoding = dynd::string_encoding_utf_8;
+      }
+      break;
+    }
+    break;
+  case 6:
+    switch (encoding_str[4]) {
+    case '1':
+      if (encoding_str == "utf_16" || encoding_str == "utf-16") {
+        encoding = dynd::string_encoding_utf_16;
+      }
+      break;
+    case '3':
+      if (encoding_str == "utf_32" || encoding_str == "utf-32") {
+        encoding = dynd::string_encoding_utf_32;
+      }
+      break;
+    }
+  }
 
-PYDYND_API PyObject *_type_array_property_names(const dynd::ndt::type &d);
+  if (encoding != dynd::string_encoding_invalid) {
+    return encoding;
+  }
+  else {
+    std::stringstream ss;
+    ss << "invalid input \"" << encoding_str << "\" for string encoding";
+    throw std::runtime_error(ss.str());
+  }
+}
 
-PYDYND_API void init_type_functions();
+inline dynd::ndt::type dynd_make_convert_type(const dynd::ndt::type &to_tp,
+                                               const dynd::ndt::type &from_tp)
```

Comment:
All of these make_*_type functions really have no purpose, they should be in Cython.